cpp/include/raft/distance/detail/pairwise_matrix/dispatch.cuh

```diff
@@ -0,0 +1,208 @@
+  // without causing misalignent errors.
+  int vec_len_aligned = (byte_alignment % sizeof(DataT) == 0) ? byte_alignment / sizeof(DataT) : 1;
+
+  dispatch(is_row_major, vec_len_aligned, [&](auto row_major, auto vec_len_aligned) {
+    // row_major and vec_len are std::integral_constants of type bool and int
+    // respectively.
+
+    // To keep compile times in check, we only specialize on veclen > 1 when
+    // the inner loop is relatively cheap (< 5 flops).
+    constexpr int vec_len_op = distance_op.expensive_inner_loop ? 1 : vec_len_aligned();
+
+    // Prevent double, vec_len=4 combination (this is not supported)
+    constexpr int vec_len = std::min(vec_len_op, static_cast<int>(16 / sizeof(DataT)));
+
+    typedef typename raft::linalg::Policy4x4<DataT, vec_len>::Policy RowPolicy;
+    typedef typename raft::linalg::Policy4x4<DataT, vec_len>::ColPolicy ColPolicy;
+    typedef typename std::conditional<row_major(), RowPolicy, ColPolicy>::type Policy;
+
+    return pairwise_matrix<Policy, row_major(), DataT, AccT, OutT, IdxT, OpT, FinOpT>(
+      distance_op, fin_op, x, y, x_norm, y_norm, m, n, k, ldx, ldy, ld_out, out, stream);
+  });
+}
+
+template <typename opT,
+          typename DataT,
+          typename AccT,
+          typename OutT,
+          typename FinOpT,
+          typename IdxT = int>
+void distance_matrix_cutlass_dispatch(opT cutlass_op,
+                                      IdxT m,
+                                      IdxT n,
+                                      IdxT k,
+                                      const DataT* x,
+                                      const DataT* y,
+                                      const DataT* x_norm,
+                                      const DataT* y_norm,
+                                      OutT* out,
+                                      FinOpT fin_op,
+                                      cudaStream_t stream,
+                                      bool is_row_major)
+{
+  // Determine leading dimensions and possibly flip order of passing x and y if
+  // column_major.
+  IdxT ldx, ldy, ld_out;
+  if (is_row_major) {
+    ldx = k, ldy = k, ld_out = n;
+  } else {
+    std::swap<const DataT*>(x, y);
+    std::swap<const DataT*>(x_norm, y_norm);
+    std::swap(m, n);
+    ldx = m, ldy = n, ld_out = n;
+  }
+
+  size_t align_x        = alignment_of_2d_array(x, ldx);
+  size_t align_y        = alignment_of_2d_array(y, ldy);
+  size_t byte_alignment = min(align_x, align_y);
+
+  // Since alignment is in bytes, it could be smaller than sizeof(DataT).
+  // Handle this (unlikely) case here.
+  RAFT_EXPECTS(sizeof(DataT) <= byte_alignment,
+               "Input matrix must be aligned to size of elements.");
+
+  // Compute number of elements that can be loaded in one instruction
+  // without causing misalignent errors.
+  int vec_len_aligned = (byte_alignment % sizeof(DataT) == 0) ? byte_alignment / sizeof(DataT) : 1;
```

Comment:
Shall we have a helper function to avoid repetition?

Reply:
Sure! The header is split up to improve compile times in PR #1307 . I will pick it up there.